users/user-guide-query/multi-stage-query/operator-types/mailbox-receive.md

```diff
@@ -0,0 +1,110 @@
+---
+description: >-
+  Describes the mailbox receive operator in the multi-stage query engine.
+---
+
+# Mailbox receive operator
+
+The mailbox receive operator is the operator that receives the data from the mailbox send operator.
+This is not an actual relational operator but a Pinot extension used to receive data from other stages.
+
+## Implementation details
+
+Stages in the multi-stage query engine are executed in parallel by different workers.
+Workers send data to each other using mailboxes.
+The number of mailboxes depends on the send operator parallelism, the receive operator parallelism and the distribution
+being used.
+At worse, there is one mailbox per worker pair, so if the upstream send operator has a parallelism of S and the receive 
+operator has a parallelism of R, there will be S * R mailboxes.
+
+By default, these mailboxes are GRPC channels, but when both workers are in the same server, they can use shared memory
+and therefore a more efficient on heap mailbox is used.
+
+The mailbox receive operator pulls data from these mailboxes and sends it to the downstream operator.
+
+### Blocking nature
+
+The mailbox receive operator is a streaming operator.
+It emits the blocks of rows as soon as they are received from the upstream operator.
+
+It is important to notice that the mailbox receive operator tries to be fair when reading from multiple workers.
+
+## Hints
+None
+
+## Stats
+### executionTimeMs
+Type: Long
+
+The summation of time spent by all threads executing the operator.
+This means that the wall time spent in the operation may be smaller that this value if the parallelism is larger than 1.
+
+### emittedRows
+Type: Long
+
+The number of groups emitted by the operator.
+This operator should always emit as many rows as its upstream operator.
+
+### fanIn
+Type: Long
+
+How many workers are sending data to this operator.
+
+### inMemoryMessages
+Type: Long
+
+How many messages have been received in heap format by this mailbox.
```

Comment:
who decides whether to send in heap vs raw format?

Reply:
Blocks are sent using mailboxes from workers executing a stage to workers executing the parent stage. If and only if the sender and receiver workers are on the same process, then the block can be sent in memory format (which means we just send a reference to the heap, which is super cheap).

So the decision is not actually on whether we should use raw or heap but on whether workers need to send data to different processes and that depends on the query, the hints, the data distribution, etc.

Basically the reason why joins are faster when they are co-located is because in co-located joins it is guaranteed that even we have several workers executing the join in parallel, we known for sure that the children workers that send data to these workers are always in the same process and therefore all blocks are sent in heap mode*.

\* Obviously after this join there will be a reduce phase that will need to serialize data to a different process, but we assume the data that is going to be sent there is going to be smaller than the one used to calculate the join